### PR TITLE
Rewrite fusion-core api to use dependency injection

### LINKED ISSUE #52: Add support for dependency injection in plugins
NOTE: This issue needs to be updated with much more detail. This is a starting point for reference

```js
// Basic example of main
// src/main.js
import App from 'fusion-core';
import PluginA from 'fusion-plugin-a';
import PluginB from 'fusion-plugin-b';
import {TokenA, TokenB} from 'fusion-types'
export default function main() {
  const app = new App();
  app.register(PluginA, TokenA);
  app.register(PluginB, TokenB);
  return app;
}
```

```js
// basic example of a plugin with dependencies
// fusion-plugin-a
import {withDependencies} from 'fusion-core';
import {TokenB} from 'fusion-types';

class MyThing {
  // some api here...
}

export default withDependencies({b: TokenB})(deps => {
  const {b} = deps;
  return new MyThing(b);
});
```

```js
// basic example of a plugin with dependencies and a middleware
// fusion-plugin-a
import {withDependencies, withMiddleware} from 'fusion-core';
import {TokenB} from 'fusion-types';

export default withDependencies({b: TokenB})(deps => {
  const {b} = deps;
  const thing = new MyThing(b);
  return withMiddleware((ctx, next) => {
    // do middleware things
    return next();
  }, thing);
});
```

```js
// example of main with middleware
import App from 'fusion-core';
import PluginA from 'fusion-plugin-a';
import PluginB from 'fusion-plugin-b';
import {TokenA, TokenB} from 'fusion-types'
export default function main() {
  const app = new App();
  app.register(PluginA, TokenA);
  app.register(PluginB, TokenB);
  
  app.middleware((deps) => (ctx, next) => {
    // use your dependencies
    deps.a.thing()
    return next();
  }, {a: TokenA});
  return app;
}
```

### PR DESCRIPTION
**Pull request by ganemone**
Fixes #52

### CODE REVIEW

File: flow-typed/fusion-core.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  declare type MemoizeFn<A> = (ctx: Context) => A;
+  declare export function memoize<A>(fn: MemoizeFn<A>): MemoizeFn<A>;
+
+  declare class FusionApp {
+    // TODO: More specific types
+    constructor<Element>(element: Element, render: (Element) => any): FusionApp;
+    registered: Map<any, any>;
+    plugins: Array<any>;
+    renderer: any;
+    // register(middleware: MiddlewarePlugin): void;
```
Comment (AlexMSmithCA) [inline flow-typed/fusion-core.js]: Are we keeping this in?

File: flow.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/* @flow */
```
Comment (AlexMSmithCA) [inline flow.js]: If these definitions are only used within `fusion-core`, what are your thoughts on not adding it to the the `[lib]` config and instead importing the specific types in the files that need it?  Benefit of not polluting the global namespace and avoiding any collision issues down the road.  Downside is more imports...

File: src/__tests__/__node__/index.js [hunk 3]
```diff
@@ -1,274 +1,182 @@
...
+        'object',
+        'ctx.template.htmlAttrs'
+      );
+      t.ok(ctx.template.head instanceof Array, 'ctx.template.head');
+      t.ok(ctx.template.body instanceof Array, 'ctx.template.body');
+      await next();
+      t.equals(
+        typeof ctx.template,
+        'object',
+        'ctx.templatekeeps structure on upstream'
```
Comment (AlexMSmithCA) [inline src/__tests__/__node__/index.js]: _(minor)_ spacing

File: flow-typed/fusion-core.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  declare export type SSRContext = {
+    element: any,
+    template: {
+      htmlAttrs: Object,
+      title: string,
+      head: Array<string>,
+      body: Array<string>,
+    },
+  } & KoaContext;
+  declare export type Context = SSRContext | KoaContext;
```
[the remark this answers is not in the document]
Comment (ganemone) [reply flow-typed/fusion-core.js]: Yea, I think that is the current plan.

File: flow-typed/fusion-core.js (hunk 1 above)
Comment (ganemone) [reply flow-typed/fusion-core.js]: Still massaging the types a bit

File: flow.js (hunk 2 above)
Comment (ganemone) [reply flow.js]: I was having trouble with that approach combined with method overrides

File: src/__tests__/dependency-resolution.js [hunk 5]
```diff
@@ -0,0 +1,286 @@
+/* @flow */
+import tape from 'tape-cup';
+import ClientAppFactory from '../client-app';
+import ServerAppFactory from '../server-app';
+import {withDependencies} from '../with-dependencies';
+import {withMiddleware} from '../with-middleware';
+
+function createToken(name) {
```
Comment (lhorie) [inline src/__tests__/dependency-resolution.js]: Shouldn't `createToken` be exposed as part of the core API?

File: scratch.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+```js
+// src/main.js
+import ConsoleLoggerPlugin from 'fusion-plugin-console-logger';
+import {LoggerToken} from 'fusion-tokens';
+import App from 'fusion-core';
+
+export default function main() {
+  const app = new App(...);
+  app.register(ConsoleLoggerPlugin, LoggerToken);
```
Comment (lhorie) [inline scratch.md]: Should document `app.configure`

File: scratch.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+In order to use plugins, you need to register them with your FusionJS application. You do this by calling
+`app.register` with the plugin and a token for that plugin. The token is simply a value used to keep track of
+what plugins are registered, and to allow plugins to depend on one another. You can think of Tokens like interfaces.
+We keep a list of standard tokens in the `fusion-tokens` repository. Lets finish up this logger example:
+
+```js
+// src/main.js
+import ConsoleLoggerPlugin from 'fusion-plugin-console-logger';
+import {LoggerToken} from 'fusion-tokens';
```
Comment (lhorie) [inline scratch.md]: We should probably make a list of what's going to live here

File: src/plugins/ssr.js [hunk 8]
```diff
@@ -105,6 +14,60 @@ function isSSR(ctx) {
...
+
+    const template = {
+      htmlAttrs: {},
+      title: '',
+      head: [],
+      body: [],
+    };
+    ctx.element = element;
+    ctx.rendered = '';
+    ctx.template = template;
```
Comment (lhorie) [inline src/plugins/ssr.js]: Unrelated breaking change? Needs migration docs

File: src/__tests__/dependency-resolution.js (hunk 5 above)
Comment (ganemone) [reply src/__tests__/dependency-resolution.js]: I was planning on having this separated into a `fusion-tokens` library and have it colocated with the default tokens

File: src/base-app.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        return resolved.get(token);
+      }
+      // if currently resolving the same type, we have a circular dependency
+      if (resolving.has(token)) {
+        throw new Error(
+          `Cannot resolve circular dependency: ${token.toString()}`
+        );
+      }
+      // the type was never registered, throw error
+      if (!registered.has(token)) {
```
Comment (AlexMSmithCA) [inline src/base-app.js]: What is the rationale for not hard failing here?

File: src/client-app.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      this.resolve();
+      const middleware = compose(this.plugins);
+      return () => {
+        // TODO: Create noop context object to match server api
+        const ctx = {
+          url: window.location.path + window.location.search,
+          element: null,
+          body: null,
+        };
+        // $FlowFixMe
```
Comment (AlexMSmithCA) [inline src/client-app.js]: _(minor)_ For flow errors we want to suppress but don't have a plan to 'fix', I'd vote we use `$FlowIgnore` as we do in other plugins.  Helpful for differentiating between what we want fixed and what we want ignored.

Can be enabled by adding the following `.flowconfig`:
```
...
[options]
suppress_comment= \\(.\\|\n\\)*\\$FlowIgnore
...
```

File: src/base-app.js (hunk 9 above)
Comment (ganemone) [reply src/base-app.js]: Supporting optional dependencies with defaults. Basically I have this `fusion-tokens` repo which has two helper exports and then a bunch of agreed upon tokens. The helpers look like this:

```
export const createToken = name => () => {
  throw new Error(`Missing required value for token: ${name}.`);
};
export const createOptionalToken = (name, defaultValue) => defaultValue;
```

This pattern allows us to enforce that if you want to define an optional parameter, you **must** provide a default value for it. Additionally, if you have a required parameter, you get nice error messages.

Comment (lhorie) [reply src/base-app.js]: Are there tests for `createOptionalToken`?

File: src/base-app.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        // Attempt to get default value
+        // NOTE: As of now, default values can only be configuration, not full blown plugins
+        this.configure(token, token());
```
Comment (ganemone) [inline src/base-app.js]: @lhorie  This is where we handle optional tokens. If the token is not optional, an error will be thrown here with a nice error message.

**Comment by lhorie** [comment]
Needs to change merge target from master to a branch
